File: src/app_setup/app_global.dart
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:logger/logger.dart';
import 'package:one_rpapp_service/one_rpapp_service.dart';
import 'package:pool/pool.dart';
import 'package:shared_preferences/shared_preferences.dart';

class Global {
  static final GlobalKey<ScaffoldState> appScaffoldKey =
      GlobalKey<ScaffoldState>();

  static final logger = Logger();

  // static final paymentInvoker =
  //     "http://192.168.29.107:8080/saragrocery/us-central1/customFunctions/generateTxnToken";

  // static final paymentInvoker =
  //     "${AppSettings.setting.payTmEndpoint}/generateTxnToken";

  static const String orderProgressKey = 'OrderProgress';
  static DeviceInfoPlugin deviceInfo = DeviceInfoPlugin();
  static String deviceDetails =
      'android_version/device_model/is_physical/(height x width)';

  static Future<void> setDeviceDetails(BuildContext context) async {
    var details = "";
    if (Platform.isAndroid) {
      final size = Get.size;

      await deviceInfo.androidInfo.then((androidInfo) {
        details += "${androidInfo.version.release}/";

        details += "${androidInfo.model}/";

        details += androidInfo.isPhysicalDevice ? 'true' : 'false';

        details +=
            "(${size.height.toStringAsFixed(0)} x ${size.width.toStringAsFixed(0)})";
        if (kDebugMode) {
          print("log");
        }
      }).catchError((error, stackTrace) {
        Global.logger.e("Unable to get device information");
      });
    }
    deviceDetails = details;
  }

  static Future<void> setOrderProgress(bool value) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      // Update the lastCache timestamp in local
      await prefs.setBool(orderProgressKey, value);
    } catch (e) {
      return Future.error(
          "unable to set order progress, error: ${e.toString()}");
    }
    return;
  }

  static final Pool pool = Pool(2, timeout: const Duration(seconds: 60));

  static Future<bool> isOrderProgress() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      var val = prefs.getBool(orderProgressKey);

      if (val == null) {
        return false;
      }
      return val;
    } catch (e) {
      return Future.error("get order progress, error: ${e.toString()}");
    }
  }
}

File: src/app_setup/app_global.ts
// lib/global.ts
import { logger } from '@/app/globalProvider'; // or any logger you prefer
import Cookies from 'js-cookie';
import { ReactNode, createContext } from 'react';

// Device info types
interface DeviceInfo {
  os: string;
  version: string;
  model: string;
  isPhysicalDevice: boolean;
  screenSize: string;
}

class Global {
  // Logger instance
  static readonly logger = logger;

  // Constants
  static readonly orderProgressKey = 'OrderProgress';
  
  // Device info
  static deviceInfo: DeviceInfo = {
    os: '',
    version: '',
    model: '',
    isPhysicalDevice: true,
    screenSize: '0 x 0'
  };

  // Pool implementation (using semaphore-like functionality)
  private static activeRequests = 0;
  private static maxConcurrent = 2;
  private static requestQueue: (() => void)[] = [];

  // Initialize device details
  static async setDeviceDetails(): Promise<void> {
    let details = '';
    
    // Detect browser/device info
    const userAgent = navigator.userAgent;
    const isMobile = /Mobi|Android|iPhone/i.test(userAgent);
    
    // Basic device detection
    let os = 'unknown';
    if (/Android/i.test(userAgent)) os = 'Android';
    if (/iPhone|iPad|iPod/i.test(userAgent)) os = 'iOS';
    if (/Windows/i.test(userAgent)) os = 'Windows';
    if (/Mac/i.test(userAgent)) os = 'Mac';

    details += `${os}/`;
    details += `${userAgent.split(' ')[0]}/`; // Simplified model
    details += isMobile ? 'true' : 'false';
    details += `(${window.innerHeight} x ${window.innerWidth})`;

    this.deviceInfo = {
      os,
      version: '', // Would need a library for proper version detection
      model: userAgent,
      isPhysicalDevice: isMobile,
      screenSize: `${window.innerHeight} x ${window.innerWidth}`
    };

    if (process.env.NODE_ENV === 'development') {
      console.log('Device details:', details);
    }
  }

  // Order progress methods
  static async setOrderProgress(value: boolean): Promise<void> {
    try {
      // Using cookies (alternative: localStorage)
      Cookies.set(this.orderProgressKey, value.toString(), { expires: 1 }); // 1 day expiry
    } catch (e) {
      this.logger.error(`Unable to set order progress: ${e}`);
      throw new Error(`Unable to set order progress: ${e}`);
    }
  }

  static async isOrderProgress(): Promise<boolean> {
    try {
      const value = Cookies.get(this.orderProgressKey);
      return value === 'true';
    } catch (e) {
      this.logger.error(`Error getting order progress: ${e}`);
      return false;
    }
  }

  // Pool implementation for concurrency control
  static async withPool<T>(fn: () => Promise<T>): Promise<T> {
    if (this.activeRequests >= this.maxConcurrent) {
      await new Promise<void>(resolve => {
        this.requestQueue.push(resolve);
      });
    }

    this.activeRequests++;
    try {
      return await fn();
    } finally {
      this.activeRequests--;
      if (this.requestQueue.length > 0) {
        const next = this.requestQueue.shift();
        next?.();
      }
    }
  }
}

// React context for global state
export const GlobalContext = createContext({
  deviceInfo: Global.deviceInfo,
  setDeviceDetails: Global.setDeviceDetails,
  setOrderProgress: Global.setOrderProgress,
  isOrderProgress: Global.isOrderProgress
});

export default Global;